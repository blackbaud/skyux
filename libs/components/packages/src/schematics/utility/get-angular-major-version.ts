import fs from 'node:fs';
import { createRequire } from 'node:module';

/**
 * Get the currently installed version of Angular.
 * Use `require` to satisfy Jest when importing from @angular/cli ESM module.
 */
export function getAngularMajorVersion(): string {
  const require = createRequire(__filename);

  try {
    const angularCliPackagePath = require.resolve('@angular/cli/package.json');
    const angularCliPackage = JSON.parse(
      fs.readFileSync(angularCliPackagePath, 'utf8'),
    );

    return angularCliPackage.version.split('.')[0];
  } catch {
    throw new Error(
      'Unable to determine Angular CLI version. Please ensure @angular/cli is installed.',
    );
  }
}
