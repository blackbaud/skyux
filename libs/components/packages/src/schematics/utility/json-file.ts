/* istanbul ignore file */

/**
 * The contents of this file were copied from '@schematics/angular'.
 * @see https://github.com/angular/angular-cli/blob/master/packages/schematics/angular/utility/json-file.ts
 */
import { JsonValue } from '@angular-devkit/core';
import { Tree } from '@angular-devkit/schematics';

import {
  Node,
  ParseError,
  applyEdits,
  findNodeAtLocation,
  getNodeValue,
  modify,
  parseTree,
  printParseErrorCode,
} from 'jsonc-parser';

export type InsertionIndex = (properties: string[]) => number;
export type JsonPath = (string | number)[];

/**
 * Handles JSONC files (JSON files that include comments).
 * @internal
 */
export class JsonFile {
  public content: string;

  private get JsonAst(): Node | undefined {
    if (this._jsonAst) {
      return this._jsonAst;
    }

    const errors: ParseError[] = [];
    this._jsonAst = parseTree(this.content, errors, {
      allowTrailingComma: true,
    });
    if (errors.length) {
      const { error, offset } = errors[0];
      throw new Error(
        `Failed to parse "${
          this.path
        }" as JSON AST Object. ${printParseErrorCode(
          error
        )} at location: ${offset}.`
      );
    }

    return this._jsonAst;
  }

  private _jsonAst: Node | undefined;

  constructor(private readonly host: Tree, private readonly path: string) {
    const buffer = this.host.read(this.path);
    if (buffer) {
      this.content = buffer.toString();
    } else {
      throw new Error(`Could not read '${path}'.`);
    }
  }

  public get(jsonPath: JsonPath): any {
    const jsonAstNode = this.JsonAst;
    if (!jsonAstNode) {
      return undefined;
    }

    if (jsonPath.length === 0) {
      return getNodeValue(jsonAstNode);
    }

    const node = findNodeAtLocation(jsonAstNode, jsonPath);

    return node === undefined ? undefined : getNodeValue(node);
  }

  public modify(
    jsonPath: JsonPath,
    value: JsonValue | undefined,
    insertInOrder?: InsertionIndex | false
  ): void {
    let getInsertionIndex: InsertionIndex | undefined;
    if (insertInOrder === undefined) {
      const property = jsonPath.slice(-1)[0];
      getInsertionIndex = (properties) =>
        [...properties, property].sort().findIndex((p) => p === property);
    } else if (insertInOrder !== false) {
      getInsertionIndex = insertInOrder;
    }

    const edits = modify(this.content, jsonPath, value, {
      getInsertionIndex,
      formattingOptions: {
        insertSpaces: true,
        tabSize: 2,
      },
    });

    this.content = applyEdits(this.content, edits);
    this.host.overwrite(this.path, this.content);
    this._jsonAst = undefined;
  }

  public remove(jsonPath: JsonPath): void {
    if (this.get(jsonPath) !== undefined) {
      this.modify(jsonPath, undefined);
    }
  }
}
