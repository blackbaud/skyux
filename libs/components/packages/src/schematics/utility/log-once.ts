import { JsonObject } from '@angular-devkit/core';
import { LogLevel } from '@angular-devkit/core/src/logger/logger';
import { SchematicContext } from '@angular-devkit/schematics';

const loggedMessages: Set<string> = new Set<string>();

export function logOnce(
  context: SchematicContext,
  level: LogLevel,
  message: string,
  metadata: JsonObject = {},
): void {
  if (!loggedMessages.has(message)) {
    context.logger.log(level, message, metadata);
    loggedMessages.add(message);
  }
}
