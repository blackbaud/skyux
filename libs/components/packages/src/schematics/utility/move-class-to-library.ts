import { Tree } from '@angular-devkit/schematics';
import { findNodes } from '@angular/cdk/schematics';
import ts from '@schematics/angular/third_party/github.com/Microsoft/TypeScript/lib/typescript';

export interface MoveClassToLibraryOptions {
  classNames: string[];
  previousLibrary: string;
  newLibrary: string;
}

export function moveClassToLibrary(
  tree: Tree,
  projectPath: string,
  sourceFile: ts.SourceFile,
  content: string,
  options: MoveClassToLibraryOptions,
): void {
  const recorder = tree.beginUpdate(projectPath);

  const indicatorsImports = findNodes(
    sourceFile,
    ts.SyntaxKind.ImportDeclaration,
  ).filter(
    (node): node is ts.ImportDeclaration =>
      ts.isImportDeclaration(node) &&
      node.moduleSpecifier.getText() === `'${options.previousLibrary}'`,
  )[0];

  if (indicatorsImports) {
    const replacedImports: { [key: string]: ts.ImportSpecifier } = {};

    const indicatorsImportStart = indicatorsImports.getStart(sourceFile);
    const indicatorsImportWidth = indicatorsImports.getWidth(sourceFile);

    options.classNames.forEach((importName) => {
      const namedImportSpecifier = findNodes(
        sourceFile,
        ts.SyntaxKind.ImportSpecifier,
      ).filter(
        (node): node is ts.ImportSpecifier =>
          ts.isImportSpecifier(node) && node.name.text === importName,
      )[0];

      if (namedImportSpecifier) {
        replacedImports[importName] = namedImportSpecifier;
      }
    });

    if (Object.keys(replacedImports).length > 0) {
      if (
        indicatorsImports.importClause?.namedBindings &&
        ts.isNamedImports(indicatorsImports.importClause.namedBindings) &&
        indicatorsImports.importClause.namedBindings.elements.length >
          Object.keys(replacedImports).length
      ) {
        Object.keys(replacedImports).forEach((importName) => {
          const namedImportStart =
            replacedImports[importName].getStart(sourceFile);
          const namedImportWidth =
            replacedImports[importName].getWidth(sourceFile);

          recorder.remove(namedImportStart, namedImportWidth);

          if (content.charAt(namedImportStart + namedImportWidth) === ',') {
            recorder.remove(namedImportStart + namedImportWidth, 1);
          }
        });
      } else {
        recorder.remove(indicatorsImportStart, indicatorsImportWidth);
      }

      recorder.insertLeft(
        indicatorsImportStart,
        `import { ${Object.keys(replacedImports).join(', ')} } from '${options.newLibrary}';\n`,
      );
    }

    tree.commitUpdate(recorder);
  }
}
