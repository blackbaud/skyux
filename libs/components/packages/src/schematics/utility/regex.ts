/**
 * Escapes a string value to be used in a `RegExp` constructor.
 * @see https://stackoverflow.com/questions/3446170/escape-string-for-use-in-javascript-regex
 */
export function regexEscape(value: string): string {
  return value.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}
