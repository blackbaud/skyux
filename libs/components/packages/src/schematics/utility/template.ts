import { UpdateRecorder } from '@angular-devkit/schematics';

import { parseFragment } from 'parse5';
import type { DefaultTreeAdapterTypes, Token } from 'parse5';

export type ElementWithLocation = DefaultTreeAdapterTypes.Element & {
  sourceCodeLocation: Required<Token.ElementLocation>;
};
export type ParentNode = DefaultTreeAdapterTypes.ParentNode & {
  sourceCodeLocation: Required<Token.ElementLocation>;
};
export type SwapTagCallback<T extends string> = (
  position: 'open' | 'close',
  tag: T,
  node: ElementWithLocation,
  content: string,
) => string;
export type SwapAttributeCallback<T extends string, U extends string> = (
  oldAttribute: T,
  newAttribute: U,
  node: ElementWithLocation,
  content: string,
) => string | null;

function isElement(node: unknown): node is ElementWithLocation {
  return !!node && typeof node === 'object' && 'tagName' in node;
}

export function isParentNode(node: unknown): node is ParentNode {
  return !!node && typeof node === 'object' && 'childNodes' in node;
}

export function getElementsByTagName(
  tagName: string,
  node: Element | ParentNode,
): ElementWithLocation[] {
  const elements: ElementWithLocation[] = [];
  const nodeQueue: (Element | ParentNode)[] = [node];

  while (nodeQueue.length) {
    const currentNode = nodeQueue.shift();
    if (
      isElement(currentNode) &&
      currentNode.tagName.toLowerCase() === tagName.toLowerCase()
    ) {
      elements.push(currentNode);
    }
    if (isParentNode(currentNode)) {
      nodeQueue.push(
        ...currentNode.childNodes.filter(
          (node: any): node is ElementWithLocation => isElement(node),
        ),
      );
    }
  }

  return elements;
}

export function parseTemplate(template: string): ParentNode {
  const document = parseFragment(template, {
    sourceCodeLocationInfo: true,
  });
  if (isParentNode(document)) {
    return document;
  }
  return parseFragment('', {
    sourceCodeLocationInfo: true,
  }) as ParentNode;
}

export function swapTags<T extends string>(
  content: string,
  recorder: UpdateRecorder,
  offset: number,
  oldTags: T[],
  callback: SwapTagCallback<T>,
  ...node: (ElementWithLocation | ParentNode)[]
): void {
  const nodeQueue = [...node];
  while (nodeQueue.length) {
    const currentNode = nodeQueue.shift();
    if (isElement(currentNode)) {
      const tagName = currentNode.tagName.toLowerCase() as T;
      if (oldTags.includes(tagName)) {
        /* istanbul ignore if */
        if (typeof currentNode.sourceCodeLocation?.startTag !== 'object') {
          // Skip if source code location is not available
          continue;
        }

        recorder.remove(
          offset + currentNode.sourceCodeLocation.startTag.startOffset,
          currentNode.sourceCodeLocation.startTag.endOffset -
            currentNode.sourceCodeLocation.startTag.startOffset,
        );
        recorder.insertLeft(
          offset + currentNode.sourceCodeLocation.startTag.startOffset,
          callback('open', tagName, currentNode, content),
        );
        if (currentNode.sourceCodeLocation.endTag?.startOffset) {
          recorder.remove(
            offset + currentNode.sourceCodeLocation.endTag.startOffset,
            currentNode.sourceCodeLocation.endTag.endOffset -
              currentNode.sourceCodeLocation.endTag.startOffset,
          );
          recorder.insertLeft(
            offset + currentNode.sourceCodeLocation.endTag.startOffset,
            callback('close', tagName, currentNode, content),
          );
        }
      }
    }
    if (
      // Only traverse child nodes if the current node is not a self-closing tag
      (!isElement(currentNode) ||
        currentNode?.sourceCodeLocation?.endTag?.startOffset) &&
      isParentNode(currentNode)
    ) {
      nodeQueue.push(...currentNode.childNodes.filter(isElement));
    }
  }
}

function safeSubstring(content: string, start: number, end: number): string {
  return start < end && start < content.length && end < content.length
    ? content.substring(start, end)
    : '';
}

export function getAttributeValueText<T extends string>(
  content: string,
  node: ElementWithLocation,
  attribute: T,
): string {
  return safeSubstring(
    content,
    node.sourceCodeLocation.attrs[attribute.toLowerCase()].startOffset +
      attribute.length,
    node.sourceCodeLocation.attrs[attribute.toLowerCase()].endOffset,
  );
}

/**
 * Swaps attributes in an HTML element based on a mapping of old attribute names to new attribute names.
 *
 * The function loops over each attribute in the node:
 * - If the attribute is in the mapping, it calls the optional callback function
 * - If the callback result is truthy, it adds the result to the output
 * - If the attribute is not in the mapping, it copies the attribute as-is
 */
export function swapAttributes<T extends string, U extends string>(
  node: ElementWithLocation,
  attributeSwaps: Record<T, U>,
  content: string,
  callback?: SwapAttributeCallback<T, U>,
): string {
  if (node.attrs.length === 0) {
    return '';
  }
  const result = [
    safeSubstring(
      content,
      node.sourceCodeLocation.startTag.startOffset + node.tagName.length + 1,
      node.sourceCodeLocation.attrs[node.attrs[0].name].startOffset,
    ),
  ];
  const swap = Object.fromEntries(
    Object.keys(attributeSwaps).map((key) => [key.toLowerCase(), key]),
  );

  const swapAttributeCallback: SwapAttributeCallback<T, U> =
    callback ??
    ((oldAttribute, newAttribute, node, content): string | null =>
      newAttribute
        ? newAttribute + getAttributeValueText(content, node, oldAttribute)
        : null);

  const copySpace = (index: number): void => {
    if (index > 0) {
      result.push(
        safeSubstring(
          content,
          node.sourceCodeLocation.attrs[node.attrs[index - 1].name].endOffset,
          node.sourceCodeLocation.attrs[node.attrs[index].name].startOffset,
        ),
      );
    }
  };

  node.attrs.forEach((attr: any, index: number) => {
    if (attr.name in swap) {
      const oldAttribute = swap[attr.name] as T;
      const newAttribute = attributeSwaps[oldAttribute] as U;
      const replacement =
        newAttribute &&
        swapAttributeCallback(oldAttribute, newAttribute, node, content);
      if (replacement) {
        copySpace(index);
        result.push(replacement);
      }
    } else {
      copySpace(index);
      result.push(
        safeSubstring(
          content,
          node.sourceCodeLocation.attrs[attr.name].startOffset,
          node.sourceCodeLocation.attrs[attr.name].endOffset,
        ),
      );
    }
  });
  return result.join('');
}

/**
 * Extracts the text content from a selector within element and returns it as a string.
 * If the element wraps anything more than a simple text node, an error is thrown.
 */
export function getText(text: DefaultTreeAdapterTypes.ChildNode[]): string {
  if (text.length === 1 && text[0].nodeName === '#text') {
    return (text[0] as DefaultTreeAdapterTypes.TextNode).value.trim();
  } else if (!text || text.length === 0) {
    return '';
  } else {
    // If the result contains something other than a single text node,
    // throw an error to indicate that the title cannot be converted.
    throw new Error(
      `The element contains additional markup that cannot be processed.`,
    );
  }
}
