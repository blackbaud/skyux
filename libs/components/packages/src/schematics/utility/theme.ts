export function getThemeStylesheets(): string[] {
  return [
    'node_modules/@skyux/theme/css/sky.css',
    'node_modules/@skyux/theme/css/themes/modern/styles.css',
  ];
}

export function ensureSkyuxStylesheets(
  existingStylesheets: string[]
): string[] {
  existingStylesheets = existingStylesheets
    ? existingStylesheets.map((x) => {
        // Angular 13 prefers root-relative stylesheet URLs.
        // See: https://stackoverflow.com/a/67075714/6178885
        return x.startsWith('@skyux/') ? `node_modules/${x}` : x;
      })
    : [];
  return [...new Set(getThemeStylesheets().concat(existingStylesheets))];
}
