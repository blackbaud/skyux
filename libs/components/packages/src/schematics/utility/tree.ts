import { Tree } from '@angular-devkit/schematics';

/**
 * Returns the contents of a required file or throws an error if it doesn't exist.
 */
export function readRequiredFile(tree: Tree, filePath: string): string {
  const data = tree.exists(filePath) && tree.readText(filePath);
  if (!data) {
    throw new Error(
      `The file '${filePath}' was expected to exist but was not found.`,
    );
  }

  return data;
}
