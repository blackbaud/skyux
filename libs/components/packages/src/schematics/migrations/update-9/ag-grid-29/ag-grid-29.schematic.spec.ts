import { Tree } from '@angular-devkit/schematics';
import { SchematicTestRunner } from '@angular-devkit/schematics/testing';
import { JSONFile } from '@schematics/angular/utility/json-file';

import { createTestApp, createTestLibrary } from '../../../testing/scaffold';

describe('ag-grid-29.schematic', () => {
  const runner = new SchematicTestRunner(
    'schematics',
    require.resolve('../../migration-collection.json')
  );

  async function setupTest(
    packageJson: { dependencies?: any } = {},
    build: { options?: any } = {},
    appOrLib: 'application' | 'library' = 'application'
  ) {
    let tree: Tree;
    if (appOrLib === 'application') {
      tree = await createTestApp(runner, { projectName: 'test' });
    } else {
      tree = await createTestLibrary(runner, { projectName: 'test' });
    }
    if (packageJson['dependencies']) {
      const packageJsonFile = new JSONFile(tree, 'package.json');
      packageJsonFile.modify(['dependencies'], {
        ...(packageJsonFile.get(['dependencies']) || {}),
        ...packageJson['dependencies'],
      });
    }
    if (build['options']) {
      const angularJsonFile = new JSONFile(tree, 'angular.json');
      angularJsonFile.modify(
        ['projects', 'test', 'architect', 'build', 'options'],
        build['options']
      );
    }
    return {
      tree,
    };
  }

  it('should verify css config', async () => {
    expect.assertions(1);
    const { tree } = await setupTest(
      {
        dependencies: {
          'ag-grid-community': '0.0.0',
          'ag-grid-angular': '0.0.0',
          'ag-grid-enterprise': '0.0.0',
        },
      },
      {
        options: {
          styles: [
            'node_modules/@skyux/theme/css/sky.css',
            'ag-grid-community/dist/styles/ag-theme-alpine.css',
          ],
        },
      }
    );
    const updatedTree = await runner.runSchematic('ag-grid-29', {}, tree);
    const workspace = new JSONFile(updatedTree, '/angular.json');
    expect(
      workspace.get([
        'projects',
        'test',
        'architect',
        'build',
        'options',
        'styles',
      ])
    ).toEqual([
      'node_modules/@skyux/theme/css/sky.css',
      'node_modules/@skyux/ag-grid/css/sky-ag-grid.css',
    ]);
  });

  it('should do nothing if AG Grid is not installed', async () => {
    expect.assertions(1);
    const { tree } = await setupTest({
      dependencies: {
        other: '27.1.1',
      },
    });
    const updatedTree = await runner.runSchematic('ag-grid-29', {}, tree);
    const packageJson = new JSONFile(updatedTree, '/package.json');
    expect(packageJson.get(['dependencies', 'other'])).toEqual('27.1.1');
  });

  it('should do nothing on library build configuration', async () => {
    expect.assertions(1);
    const { tree } = await setupTest(
      {
        dependencies: {
          'ag-grid-community': '0.0.0',
          'ag-grid-angular': '0.0.0',
        },
      },
      {},
      'library'
    );
    const angularJson = new JSONFile(tree, '/angular.json');
    const beforeConfig = angularJson.get(['projects', 'test', 'architect']);
    const updatedTree = await runner.runSchematic('ag-grid-29', {}, tree);
    const updatedAngularJson = new JSONFile(updatedTree, '/angular.json');
    expect(updatedAngularJson.get(['projects', 'test', 'architect'])).toEqual(
      beforeConfig
    );
  });

  it('should update selectThisNode', async () => {
    const { tree } = await setupTest({
      dependencies: {
        'ag-grid-community': '0.0.0',
        'ag-grid-angular': '0.0.0',
      },
    });
    tree.overwrite(
      'src/app/app.component.ts',
      `
        export class AppComponent {
          public updateCheckbox(column: Column, selected: boolean) {
            if (!column.isLockVisible()) {
              return this.#row.selectThisNode(selected);
            }
          }
        }`
    );
    tree.create(
      'src/app/no-change.component.ts',
      `export class NoChangeComponent {}`
    );
    const updatedTree = await runner.runSchematic('ag-grid-29', {}, tree);
    expect(updatedTree.readText('src/app/app.component.ts'))
      .toMatchInlineSnapshot(`
      "
              export class AppComponent {
                public updateCheckbox(column: Column, selected: boolean) {
                  if (!column.getColDef().lockVisible) {
                    return this.#row.setSelected(selected);
                  }
                }
              }"
    `);
  });

  it('should update RowNode to use interface', async () => {
    const { tree } = await setupTest({
      dependencies: {
        'ag-grid-community': '0.0.0',
        'ag-grid-angular': '0.0.0',
      },
    });
    tree.overwrite(
      'src/app/app.component.ts',
      `
        import { RowNode } from 'ag-grid-community';

        export class AppComponent {
          public updateRow(row: RowNode) {
            const options = {
              event: RowNode.EVENT_CELL_CHANGED,
            };
            row.addEventListener(RowNode.EVENT_CELL_CHANGED, () => undefined)
          }
        }`
    );
    tree.create(
      'src/app/no-change.component.ts',
      `export class NoChangeComponent {}`
    );

    // The spec says sourceRoot is optional, so testing a case without it.
    const angularJson = new JSONFile(tree, '/angular.json');
    angularJson.remove(['projects', 'test', 'sourceRoot']);

    const updatedTree = await runner.runSchematic('ag-grid-29', {}, tree);
    expect(updatedTree.readText('src/app/app.component.ts'))
      .toMatchInlineSnapshot(`
      "
              import { RowNode, IRowNode } from 'ag-grid-community';

              export class AppComponent {
                public updateRow(row: IRowNode) {
                  const options = {
                    event: RowNode.EVENT_CELL_CHANGED,
                  };
                  row.addEventListener(RowNode.EVENT_CELL_CHANGED, () => undefined)
                }
              }"
    `);
  });
});
