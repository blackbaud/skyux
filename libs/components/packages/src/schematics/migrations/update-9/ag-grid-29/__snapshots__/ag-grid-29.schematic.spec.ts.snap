// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ag-grid-29.schematic should update RowNode to use interface 1`] = `
"
        import { RowNode, IRowNode } from 'ag-grid-community';

        export class AppComponent {
          public updateRow(row: IRowNode) {
            const options = {
              event: RowNode.EVENT_CELL_CHANGED,
            };
            row.addEventListener(RowNode.EVENT_CELL_CHANGED, () => undefined)
          }
        }"
`;

exports[`ag-grid-29.schematic should update selectThisNode 1`] = `
"
        export class AppComponent {
          public updateCheckbox(selected: boolean) {
            return this.#row.setSelected(selected);
          }
        }"
`;
